epsagon/wrappers/django.py

```diff
@@ -26,34 +26,51 @@ class DjangoMiddleware(object):
 
     def __init__(self, get_response):
         self.get_response = get_response
-        self.runner = None
-        self.request = None
-        self.response = None
-        self.ignored_request = False
+
         epsagon.trace.trace_factory.switch_to_multiple_traces()
 
-    def process_exception(self, _, process_exception):
-        if process_exception:
-            traceback_data = get_traceback_data_from_exception(
-                process_exception
-            )
-            self.runner.set_exception(process_exception, traceback_data, False)
+    def process_exception(self, request, process_exception):
+        if not process_exception:
+            return
+
+        if not hasattr(request, "epsagon_trace") or not request.epsagon_trace.runner:
+            return
+
+        traceback_data = get_traceback_data_from_exception(process_exception)
+
+        request.epsagon_trace.runner.set_exception(
+            process_exception,
+            traceback_data,
+            False
+        )
 
     def __call__(self, request):
+        # Link epsagon to the request object for easy-access to epsagon lib
+        request.epsagon = epsagon
+
+        if epsagon.http_filters.is_ignored_endpoint(request.path):
+            return self.get_response(request)
+
+        request.epsagon_trace = epsagon.trace.trace_factory.get_or_create_trace()
+
+        request_middleware = DjangoRequestMiddleware(request)
+        request_middleware.before_request()
```

Comment:
Looks good. Just need to keep in mind it might affect existing behavior, depending on additional middleware configured (other than epsagon) - epsagon.after_request() is moved to the end of the middleware chain, regardless of its order in the django framework.

For example, before this change, if you had 3 middlewares and epsagon was #2, then you'll have:

middleware1.before_request()
epsagonMiddleware.before_request()
middleware3.before_request()
ACTUAL REQUEST
middleware1.after_request()
**epsagon.after_request()**
middleware3.after_request()

What will happen now is:
middleware1.before_request()
epsagonMiddleware.before_request()
middleware3.before_request()
ACTUAL REQUEST
middleware1.after_request()
middleware3.after_request()
**epsagon.after_request()**